codec/src/test/java/io/netty/handler/codec/compression/ZstdDecoderTest.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2021 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.compression;
+
+import com.github.luben.zstd.ZstdOutputStream;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.embedded.EmbeddedChannel;
+import org.junit.Test;
+import org.junit.experimental.theories.DataPoints;
+
+import java.io.ByteArrayOutputStream;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+public class ZstdDecoderTest extends AbstractDecoderTest {
+
+    public ZstdDecoderTest() throws Exception {
+    }
+
+    @Override
+    public void initChannel() {
+        channel = new EmbeddedChannel(new ZstdDecoder());
+    }
+
+    @DataPoints("smallData")
+    public static ByteBuf[] smallData() {
+        ByteBuf direct = Unpooled.directBuffer(compressedBytesSmall.length);
+        direct.writeBytes(compressedBytesSmall);
+        return new ByteBuf[] {direct};
+    }
+
+    @DataPoints("largeData")
+    public static ByteBuf[] largeData() {
+        ByteBuf direct = Unpooled.directBuffer(compressedBytesLarge.length);
+        direct.writeBytes(compressedBytesLarge);
+        return new ByteBuf[] {direct};
+    }
+
+    @Test(expected = DecompressionException.class)
+    public void testUnknownTypeOfChunk() throws Exception {
+
+        ByteBuf in = Unpooled.directBuffer();
+        in.writeByte(0xFF);   //random value
+        in.writeInt(0);
+
+        channel.writeInbound(in);
+    }
+
+    public void testDecompressionOfBatchedFlow(final ByteBuf expected, final ByteBuf data) throws Exception {
+        final int compressedLength = data.readableBytes();
+        int written = 0;
+        ByteBuf compressedBuf = data.slice(written, compressedLength - written);
```

Comment:
This override eliminates the whole point of `testDecompressionOfBatchedFlow`. Your current implementation works well in a synthetic test using `EmbeddedChannel` and in-memory data. In reality, network packs data in frames. It's a frequent use-case that data arrive to the server in multiple chunks of a different length. The expected length should be taken from zstd header and decoder should wait until that data available before attempting to decompress.

I see that the api you need currently is not public in zstd library, but in exists internally. Consider opening an issue in their repo and ask for one of these 2 options:
1. Add a method to `Zstd` which will determite a full length of the compressed buffer. By knowing it, you can write a decoder which  waits untill all data available before performing decompression.
2. Make `decompressStream` methods public. Looking at `ZstdInputStreamNoFinalizer` and `ZstdDirectBufferDecompressingStream`, they have what you need to decompress data in chunks, when they arrive.

Reply:
This is indeed a very complicated part, thanks for your suggestions, but from what I understand, it is really difficult to determine the size of a compressed block when the network is transmitted in blocks. It takes more work to figure out how much data is a complete compressed message, appreciate it

Reply:
With existing API - yes, unless `Zstd.decompressedSize(...)` may be used as an indicator as described in https://github.com/netty/netty/pull/10422#discussion_r586572167.
But the streaming feature is already implemented in Zstd library, it's just private at this moment. If you will describe the use-case, they can consider making it public.

Reply:
If Zstd Library can add this api, and this api can work, this solution is possible, thanks for your suggestion